File: Node/logging-standards/src/order-service-good.js
'use strict';
/**
 * src/order-service-good.js — FULLY COMPLIANT reference implementation
 *
 * Demonstrates every logging standard enforced by the fitness function:
 *
 *   ✓ Shared pino logger factory — no console.*, no direct pino require
 *   ✓ Structured object as first argument — no template literals in log calls
 *   ✓ Static dot-namespaced event names — queryable in any log aggregator
 *   ✓ Correlation context from caller — no repetition at every log site
 *   ✓ Throw OR log — never both in the same catch block
 */

const { createLogger } = require('./logger');
const { randomUUID } = require('crypto');

const log = createLogger({ module: 'order-service' });

const orders = new Map();

/**
 * Process an order.
 *
 * @param {string} orderId
 * @param {string} userId
 * @param {Object} [reqLog] — child logger pre-bound with request context
 *   (e.g. requestId, correlationId). Created by the HTTP handler and
 *   threaded through so all log lines share the same correlation fields.
 */
function processOrder(orderId, userId, reqLog) {
  const boundLog = (reqLog || log).child({ orderId, userId });

  boundLog.info({}, 'order.processing_started');

  const order = orders.get(orderId);
  if (!order) {
    boundLog.warn({}, 'order.not_found');
    return { status: 'not_found' };
  }

  // Raise the exception — the HTTP handler will catch it and decide whether
  // to log it. No duplicate log entries.
  const result = chargePayment(order);

  boundLog.info({ status: result.status, amount: result.amount }, 'order.completed');
  return result;
}

function createOrder(userId, amount) {
  const orderId = randomUUID();
  orders.set(orderId, { orderId, userId, amount });
  log.info({ orderId, userId, amount }, 'order.created');
  return { orderId, status: 'pending', amount };
}

function chargePayment(order) {
  if (order.amount > 10_000) {
    // Raise without logging — caller decides how to handle it
    const err = new Error('Amount exceeds limit');
    err.amount = order.amount;
    throw err;
  }
  return { status: 'charged', amount: order.amount };
}

module.exports = { processOrder, createOrder };

File: .Net/logging-standards/src/OrderServiceGood.cs
using System;
using System.Collections.Generic;
using Microsoft.Extensions.Logging;

namespace FitnessFunctions.Logging.Good;

/// <summary>
/// OrderServiceGood — FULLY COMPLIANT reference implementation
///
/// Demonstrates every logging standard enforced by the fitness function:
///
///   ✓ ILogger&lt;T&gt; injected via constructor — no direct instantiation
///   ✓ Structured message templates with {Property} placeholders
///   ✓ No Console.Write / Console.WriteLine
///   ✓ No string interpolation or concatenation in log calls
///   ✓ Throw OR log — never both in the same catch block
///   ✓ Scopes used for correlation context
/// </summary>
public class OrderServiceGood
{
    private readonly ILogger<OrderServiceGood> _logger;
    private readonly Dictionary<string, Order> _orders = new();

    // ILogger<T> injected by the DI container configured with Serilog
    public OrderServiceGood(ILogger<OrderServiceGood> logger)
    {
        _logger = logger;
    }

    public Order ProcessOrder(string orderId, string userId)
    {
        // Use a logging scope to bind correlation context.
        // Every log statement within the using block carries OrderId and UserId.
        using var scope = _logger.BeginScope(new Dictionary<string, object>
        {
            ["OrderId"] = orderId,
            ["UserId"] = userId,
        });

        // Structured template — {Property} placeholders, not string interpolation
        _logger.LogInformation("order.processing_started {OrderId} {UserId}", orderId, userId);

        if (!_orders.TryGetValue(orderId, out var order))
        {
            _logger.LogWarning("order.not_found {OrderId}", orderId);
            throw new KeyNotFoundException(orderId);
        }

        // Raise without logging — caller decides how to handle the exception
        ChargePayment(order);

        _logger.LogInformation(
            "order.completed {OrderId} {Status} {Amount}",
            orderId, order.Status, order.Amount);

        return order;
    }

    public Order CreateOrder(string userId, decimal amount)
    {
        var orderId = Guid.NewGuid().ToString();
        var order = new Order(orderId, userId, amount);
        _orders[orderId] = order;

        _logger.LogInformation(
            "order.created {OrderId} {UserId} {Amount}",
            orderId, userId, amount);

        return order;
    }

    public void CancelOrder(string orderId, string reason)
    {
        using var scope = _logger.BeginScope(new Dictionary<string, object>
        {
            ["OrderId"] = orderId,
        });

        if (!_orders.TryGetValue(orderId, out var order))
        {
            _logger.LogWarning("order.not_found {OrderId}", orderId);
            throw new KeyNotFoundException(orderId);
        }

        if (order.Status is "cancelled" or "shipped")
        {
            _logger.LogWarning(
                "order.cancel_rejected {OrderId} {Status} {Reason}",
                orderId, order.Status, "terminal_state");
            throw new InvalidOperationException($"Cannot cancel order in state {order.Status}");
        }

        var previousStatus = order.Status;
        order.Status = "cancelled";
        _logger.LogInformation(
            "order.cancelled {OrderId} {PreviousStatus} {CancelReason}",
            orderId, previousStatus, reason);
    }

    private static void ChargePayment(Order order)
    {
        if (order.Amount > 10_000)
            throw new InvalidOperationException($"Amount {order.Amount} exceeds limit");

        order.Status = "charged";
    }
}

public record Order(string OrderId, string UserId, decimal Amount)
{
    public string Status { get; set; } = "pending";
}
